src/digital.rs

```diff
@@ -174,3 +174,22 @@ pub trait InputPin {
     /// Is the input pin low?
     fn try_is_low(&self) -> Result<bool, Self::Error>;
 }
+
+/// Single input pin that can switch from input to output mode, and vice-versa.
+pub trait IoPin<TInput, TOutput> 
+where TInput : InputPin + IoPin<TInput, TOutput>,
+      TOutput : OutputPin + IoPin<TInput, TOutput> {
+    /// Error type.
+    type Error;
+
+    /// Tries to convert this pin to input mode.
+    /// 
+    /// If the pin is already in input mode, this method should succeed.
+    fn try_switch_to_input_pin(self) -> Result<TInput, Self::Error>;
+
+    /// Tries to convert this pin to output mode with the given initial state.
+    /// 
+    /// If the pin is already in the requested state, this method should
+    /// succeed.
+    fn try_switch_to_output_pin(self, state: PinState) -> Result<TOutput, Self::Error>;
```

Comment:
Would _not_ setting an initial state also make sense?

Reply:
If the user didn't specify the state, then the result would be a pin in output mode that's either high or low, but the user doesn't know which it is. I think it's better to make the user specify which state they want rather than to leave it ambiguous.

That being, said I'm no expert in the embedded space. Do you see a valid use case?

Reply:
I was just wondering. From a software perspective it makes sense to always change to a defined state. However, from a hardware perspective, it may not _always_ make sense to set a pin, since it has electrical implications.
Does anybody else here have an use case for changing the pin type to output but leaving it unset?

Reply:
IIUC, an example use-case might be an output pin in open drain mode (I could be wrong; like I said, I'm not an embedded expert). However, the current digital traits don't seem to provide any support around these states either. Maybe it makes sense to look at that separately, eg. as an addition to PinState? Definitely interested in what other folks think here, as @eldruin said.